File: 프로그래머스/lv2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.LinkedList;
import java.util.Queue;
class Solution {
    static boolean bd_ck[][];
    public int solution(int m, int n, String[] board) {
        int answer = 0;
        String[][] bd = new String[m][n];
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                bd[i][j] = board[i].split("")[j];
            }
        }
        
        boolean flag = true;
        while(flag){
            bd_ck = new boolean[m][n];
            flag = false;
            for(int i=0; i<m-1; i++){
                for(int j=0; j<n-1; j++){
                    if(bd[i][j].equals("#")) continue;
                    if(check(i,j,bd)){
                        bd_ck[i][j] = true;    
                        bd_ck[i][j+1] = true;
                        bd_ck[i+1][j] = true;
                        bd_ck[i+1][j+1] = true;
                        flag = true;
                    }
                }
            }
            answer += drop(m, n, bd);
            bd_ck = new boolean[m][n];
        }
        return answer;
    }
    public static boolean check(int i, int j, String[][] bd){
        String s = bd[i][j];
        if(s.equals(bd[i][j+1]) && s.equals(bd[i+1][j]) && s.equals(bd[i+1][j+1])){
            return true;
        }
        return false;
    }

    public static int drop(int m, int n, String[][] bd){
        int cnt = 0;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(bd_ck[i][j]){
                    bd[i][j] = ".";
                }
            }
        }

        //m=4 , n=5
        for(int i=0; i<n; i++){
            Queue<String> stack = new LinkedList<>();
            for(int j=m-1; j>=0; j--){
                if(bd[j][i].equals(".")){
                    cnt++;
                }
                else{
                    stack.add(bd[j][i]);
                }
            }
            int idx = m-1;

            while(!stack.isEmpty()){
                bd[idx][i] = stack.poll();
                idx--;
            }
            
            for(int k=idx; k>=0; k--){
                bd[k][i] = "#";
            }
        }
        return cnt;
    }
}

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.js
function solution(m, n, board) {
    let answer = 0;
    
    board = board.map(v => v.split(''));
    
    const getDestroyBlock = () => {
        const arr = [];
        for (let i = 0; i < m - 1; i++) {
            for (let j = 0; j < n - 1; j++) {
                if (!board[i][j]) continue;
                if (board[i][j] === board[i + 1][j] && board[i][j] === board[i][j + 1] && board[i][j] === board[i + 1][j + 1]) {
                    arr.push([i, j], [i + 1, j], [i, j + 1], [i + 1, j + 1]);
                }
            }
        } 
        return arr;
    }
    
    const downBlock = () => {
        
        for (let i = 0; i < n; i++) {
            const stack = [];
            for (let j = m - 1; j >= 0; j--) {
                if (!board[j][i]) continue;
                stack.push(board[j][i]);
                board[j][i] = null;
            }
            
            for (let j = 0; j < stack.length; j++) {
                board[m - j - 1][i] = stack[j];
            }
        }
    }
    
    while (true) {
        const results = getDestroyBlock();

        if (!results.length) break;
        
        for (const [y, x] of results) {
            board[y][x] = null;
        }
        
        downBlock();
    }
    
    return board.flat().filter(v => v === null).length;
}